### PR TITLE
Add :tabhistory command

### LINKED ISSUE #242: Tab history (back/forward) list
The Firefox [history](https://developer.mozilla.org/en-US/docs/Web/API/History) API doesn't allow a WebExt (or webpage) to see the tab history list. You can only go back by a count (e.g. `history.back() === history.go(-1) === Tridactyl "H" nmap`). Lame.

However, Tridactyl is more betterer than dumb page JS, and it could maintain its own per-tab forward/back history list which can be used in a completion-list for `back` and `forward`.

It would probably take a little bit of finagling to ensure it keeps in sync with the FF version accessed by right-clicking on the back/forward arrows, but I _think_ it would be possible.

### PR DESCRIPTION
**Pull request by ivcz**
Closes #242 

There probably are better ways to get favicons, but I couldn't figure it out

### CODE REVIEW

**Comment by bovine3dom** [comment]
Thanks for this PR! I'm travelling at the moment so it might take a while to get round to reviewing it.

If I haven't said anything in a couple of weeks please poke me because I've probably forgotten :)

**Comment by ivcz** [comment]
Bump
Also I just noticed that I pushed into master instead of a feature branch in my repo is it okay?

**Comment by bovine3dom** [comment]
Thanks for the reminder : ) I'll try to get to this by the end of next week.

It doesn't really matter that it's your master branch, so long as you don't push unrelated things to it by accident before I merge it.

**Comment by bovine3dom** [comment]
I swapped your `:tabhistory` for `:back` and `:forward` because it made more sense to me.

We don't seem to get history entries for Tridactyl pages (e.g. `:help`) or the website https://txti.es (and presumably others). If you get a chance, could you look into it? If not I'll have a go at some point :)

**Comment by ivcz** [comment]
Ok, I'll see what i can do

**Comment by ivcz** [comment]
As far as I can tell the problem is that DOMContentLoaded is not firing on either help pages or txti, possibly some other pages too.
I think addCurHistory() call should be placed somewhere in autocmds, but I can't figure out how to do that properly, maybe you can steer me in the right direction?

**Comment by bovine3dom** [comment]
That is to cope with webapp stuff but it's not pretty - we check ten times a second if the URI has changed on a page: https://github.com/tridactyl/tridactyl/blob/9d853d2a89a161c6dc8f3cccb3f04ff2e0872fe8/src/excmds.ts#L2043

That code only runs if an autocommand exists for that URL, so we can get away with it because most people won't have UriChange autocommands for most pages.

If we did something similar for every single page, we'd reduce battery life quite a lot because we'd stop every single tab from going to sleep.

We could potentially do it more neatly - e.g. listening for UriChanges for a short while after each interaction with the page - but we'd need to think about it quite a bit : )

**Comment by ivcz** [comment]
Got it. I'd guess updating on some key events like scroll and blur would be enough, but that's more of a blunt approach.

There is an issue with search engines and probably other sites that add get params when you revisit them, which might break indexing since the link isn't the same when you go back

**Comment by ivcz** [comment]
I'll test using document title as an index later since I can't remember why I opted to use page url in the first place, but there'll still be some weird behavior, I suppose

**Comment by ivcz** [comment]
Sounds like it, because I started with using title as an index like in the jump list
Guess I'll just reindex the sequence when it breaks

**Comment by bovine3dom** [comment]
I'm not sure. We need to make our own cache really for use with history completions on :open

It's possible to get favicons from google - https://github.com/tridactyl/tridactyl/issues/197 - but that should probably be off by default if we do include it because people won't like it

**Comment by ivcz** [comment]
I'll just remove them for now then

**Comment by bovine3dom** [comment]
Thanks for this, sorry I've left it so long! It seems to work pretty well outside of SPAs now.

I'm just thinking that it would be nice to have a timestamp for when each page was last visited. Would you be interested in adding that?

**Comment by GHolk** [comment]
Hey @ivcz and @bovine3dom ,
it has been a while since last comment on this PR.
I am looking forward to this feature.
Do you have any progress? or may I add the timestamp feature for you?

**Comment by bovine3dom** [comment]
@GHolk I was hoping to fix #3969 so we could have the history work on single-page applications, but I don't think that's likely.

I just need to review the tree logic, at the moment it displays things like

```
 | | | | | ├─
```

but I would prefer it if it was like

```
      ├──
```

If I can't work that out before the next release I'll merge it anyway as it's already working mostly fine :)

**Comment by GHolk** [comment]
I did not notice this PR implement a tree-style history; tree is a little complicate to me.
I will try adding the timestamp feature when I am available.
By the way, if we want to ensure the timestamp is same as the timestame in firefox's history,
we have to get the timestamp from WebExtension's `history.search`,
which will be a little troublesome.

File: src/excmds.ts [hunk 1]
```diff
@@ -1089,6 +1103,37 @@ document.addEventListener("scroll", addJump, { passive: true })
...
+    const link = getJumpPageId()
+    const current = pages["list"].findIndex(item => item.href === link)
+    if (pages["list"][current]) {
+        pages["current"] = current
+    } else {
+        pages["list"].push({
+            parent: pages["current"],
+            href: link,
+            title: document.title,
+            id: pages["list"].length,
```
Comment (bovine3dom) [inline src/excmds.ts:1125]: @GHolk I think we can just use store the system timestamp here, I don't think it should be too complicated

**Comment by GHolk** [comment]
A easy implementation, with a handcraft timespan  format function.
[branch tab-history](/GHolk/tridactyl/tree/tab-history)

**Comment by bovine3dom** [comment]
this is pretty much done

I need to 

- double check the behaviour of sessions in private windows
- make the tree-ness more compact so that normal browsing sessions don't become ridiculous

**Comment by bovine3dom** [comment]
LGTM, I'll merge once the tests have passed

**Comment by GHolk** [comment]
Hey, the theme shydactyl will not show up the timestamp.
It seem that remove the `:root #completions table tr .title { width: 50%; }` in shydactyl.css can fix this issue,
but Im not sure if this will cause other issues.

**Pull request merged**